pkg/daemon/irreconcilable_diff_reporter.go

```diff
@@ -0,0 +1,224 @@
+	return json.Marshal(config)
+}
+
+func (i *irreconcilableDifferencesReportGeneratorImpl) fetchBaseInstallTimeMachineConfig() *mcfgv1.MachineConfig {
+	for _, path := range i.installTimeParamsPaths {
+		content, err := os.ReadFile(path)
+		if err != nil {
+			continue
+		}
+		var mc mcfgv1.MachineConfig
+		if err := json.Unmarshal(content, &mc); err == nil {
+			return &mc
+		}
+	}
+	return nil
+}
+
+// NoOpIrreconcilableReporterImpl a no-op irreconcilable differences reporter that does nothing
+// when CheckReportIrreconcilableDifferences is called.
+type NoOpIrreconcilableReporterImpl struct{}
+
+// NewNoOpIrreconcilableReporterImpl creates a new instance of NoOpIrreconcilableReporterImpl
+func NewNoOpIrreconcilableReporterImpl() *NoOpIrreconcilableReporterImpl {
+	return &NoOpIrreconcilableReporterImpl{}
+}
+
+// CheckReportIrreconcilableDifferences does nothing when called. Returned error is always nil.
+func (r *NoOpIrreconcilableReporterImpl) CheckReportIrreconcilableDifferences(_ *mcfgv1.MachineConfig, _ string) error {
+	return nil
+}
+
+// IrreconcilableReporterImpl handles creating irreconcilable differences reports for the MCN CR and pushing them
+// on each node update.
+type IrreconcilableReporterImpl struct {
+	mcfgClient mcfgclientset.Interface
+	// irreconcilableDifferencesReportGenerator creates the diff report itself.
+	diffGenerator irreconcilableDifferencesReportGenerator
+}
+
+// newIrreconcilableReporter is the internal constructor of IrreconcilableReporterImpl that allows providing
+// a non-default irreconcilableDifferencesReportGenerator.
+func newIrreconcilableReporter(mcfgClient mcfgclientset.Interface, diffGenerator irreconcilableDifferencesReportGenerator) *IrreconcilableReporterImpl {
+	return &IrreconcilableReporterImpl{
+		mcfgClient:    mcfgClient,
+		diffGenerator: diffGenerator,
+	}
+}
+
+// NewIrreconcilableReporter creates a new instance of IrreconcilableReporterImpl.
+func NewIrreconcilableReporter(
+	mcfgClient mcfgclientset.Interface,
+) *IrreconcilableReporterImpl {
+	return newIrreconcilableReporter(
+		mcfgClient,
+		newIrreconcilableDifferencesReportGeneratorImpl(
+			[]string{
+				"/etc/mcs-machine-config-content.json",
+			},
+		),
+	)
+}
+
+func (r *IrreconcilableReporterImpl) CheckReportIrreconcilableDifferences(targetMachineConfig *mcfgv1.MachineConfig, nodeName string) error {
+	mcNode, err := r.mcfgClient.MachineconfigurationV1().MachineConfigNodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		// no existing MCN found since no resource found, no error yet just create a new one
+		if apierrors.IsNotFound(err) {
+			klog.Warningf("MCN for %s node does not exits. Skipping irreconcilable differences reporting", nodeName)
+			return nil
+		}
+		return err
+	}
+
+	diffs, err := r.diffGenerator.GenerateReport(targetMachineConfig)
+	if err != nil {
+		klog.Errorf("error generating the irreconcilable differences list of changes. %v", err)
```

Comment:
Curious why the mix of logged errors and returned errors? It looks like we would just be logging this later regardless?